lmorpho/quaternion.hpp

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <cmath>
+
+struct quaternion {
+    double w, x, y, z;
+
+    constexpr quaternion(): w(0), x(0), y(0), z(0) {}
+
+    // scalar
+    constexpr quaternion(double w): w(w), x(0), y(0), z(0) {}
+
+    // vector (pure imaginary)
+    constexpr quaternion(double x, double y, double z): w(0), x(x), y(y), z(z) {}
+
+    // all 4-components
+    constexpr quaternion(double w, double x, double y, double z): w(w), x(x), y(y), z(z) {}
+
+    quaternion conj() const {
+        return {w, -x, -y, -z};
+    }
+
+    quaternion operator*(quaternion q) const {
```

Comment:
Looking at wikipedia (so that might be wrong) I do not understand the multiplication here.